### ISSUE TITLE
Add support for internal SD card reader, a.k.a. "Baseqi Ninja disk"

### ISSUE THREAD

**Issue by OldManLink** [syscl/Fix-usb-sleep#13]
I found this project after getting the `Disk Not Ejected Properly` error for my **[Baseqi Ninja](https://shop.baseqi.com/?gclid=CjwKCAjw8O7bBRB0EiwAfbrThympI1-jJtryhndrzcYX0xbc55VIhv0E7ArdY3bYSKpwJnXvsdRYyxoCb3oQAvD_BwE)** disk. This is an internal "disk" that uses the SD card reader on my MacBook Pro. Here are the details of the disk when it is mounted: 

```/dev/disk2 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *196.9 GB   disk2
   1:                        EFI EFI                     209.7 MB   disk2s1
   2:                  Apple_HFS Baseqi Ninja            196.5 GB   disk2s2
```
Would it be possible to extend `Fix-usb-sleep` to work for this disk also? I have of course tried it, but it did not help with the problem.

**Comment by syscl** [comment]
@OldManLink Thank you for your information, in short, the answer is yes. But would you mind provide more information about your SD slot? In terminal type in 
```sh
diskutil info disk2
diskutil info disk2s1
diskutil info disk2s2
```

Thank you in advance,
syscl

**Comment by syscl** [comment]
Thank you, I will add the SD support now.

**Comment by syscl** [comment]
@OldManLink Could you please provide the log by running the script directly by typing
```sh
sudo /etc/sysclusbfix.sleep
```
Note: please insert your SD card as well.

syscl

**Comment by OldManLink** [comment]
This is what happens:
```
19:37 $ sudo /etc/sysclusbfix.sleep
Disk disk2 ejected
usage: ioreg [-abfilrtx] [-c class] [-d depth] [-k key] [-n name] [-p plane] [-w width]
where options are:
	-a archive output
	-b show object name in bold
	-c list properties of objects with the given class
	-d limit tree to the given depth
	-f enable smart formatting
	-i show object inheritance
	-k list properties of objects with the given key
	-l list properties of all objects
	-n list properties of objects with the given name
	-p traverse registry over the given plane (IOService is default)
	-r show subtrees rooted by the given criteria
	-t show location of each subtree
	-w clip output to the given line width (0 is unlimited)
	-x show data and numbers as hexadecimal
/etc/sysclusbfix.sleep: line 47: [: too many arguments
/etc/sysclusbfix.sleep: line 53: /Applications/Wireless Network Utility.app/rfoff.rtl: No such file or directory
The file /Applications/Wireless Network Utility.app does not exist.
✘-1 ~/CodeRoot/other/sleepFix [master L|✔] 
19:37 $ 
```
Now the disk icon has gone from the desktop, so it looks like it was ejected OK. However I am not sure what to do to get it back again.

**Comment by syscl** [comment]
Remount it by the following
```sh
sudo /etc/sysclusbfix.wake
```

**Comment by OldManLink** [comment]
Unfortunately that does not work:
```
19:37 $ sudo /etc/sysclusbfix.wake
Unable to find disk for disk2s2
usage: ioreg [-abfilrtx] [-c class] [-d depth] [-k key] [-n name] [-p plane] [-w width]
where options are:
	-a archive output
	-b show object name in bold
	-c list properties of objects with the given class
	-d limit tree to the given depth
	-f enable smart formatting
	-i show object inheritance
	-k list properties of objects with the given key
	-l list properties of all objects
	-n list properties of objects with the given name
	-p traverse registry over the given plane (IOService is default)
	-r show subtrees rooted by the given criteria
	-t show location of each subtree
	-w clip output to the given line width (0 is unlimited)
	-x show data and numbers as hexadecimal
/etc/sysclusbfix.wake: line 22: [: too many arguments
/etc/sysclusbfix.wake: line 28: /Applications/Wireless Network Utility.app/rfoff.rtl: No such file or directory
The file /Applications/Wireless Network Utility.app does not exist.
✘-1 ~/CodeRoot/other/sleepFix [master L|✔] 
19:42 $ 
```

**Comment by syscl** [comment]
@OldManLink It seems there's some error, please give me a cat of the sleep script by
```sh
/etc/sysclusbfix.sleep
```

syscl

**Comment by syscl** [comment]
How about 
```sh
diskutil list
```

**Comment by OldManLink** [comment]
Here you go:
```
19:42 $ cat /etc/sysclusbfix.sleep
#!/bin/sh
#
# This script aims to unmount all external devices automatically before sleep.
#
# Without this procedure, various computers with OS X/Mac OS X(even on a real Mac) suffer from "Disk not ejected properly"
# issue when therere external devices plugged-in. Thats the reason why I created this script to fix this issue. (syscl/lighting/Yating Zhou)
#
# All credit to Bernhard Baehr (bernhard.baehr@gmx.de), without his great sleepwatcher dameon, this fix will not be created.
#

#
# Added unmount Disk for "OS X" (c) syscl/lighting/Yating Zhou.
#
gMountPartition="/tmp/com.syscl.externalfix"
gDisk=($(diskutil list |grep -i -i "dev" |grep -i -o "disk[0-9]"))

for ((i=0; i<${#gDisk[@]}; ++i))
do
  gProtocol=$(diskutil info ${gDisk[i]} |grep -i "Protocol" |sed -e "s|Protocol:||" -e "s| ||g")
  if [[ ${gProtocol} == *"USB"* ]];
    then
      gCurrent_Partitions=($(diskutil list ${gDisk[i]} |grep -o "disk[0-9]s[0-9]"))
      for ((k=0; k<${#gCurrent_Partitions[@]}; ++k))
      do
        gConfirm_Mounted=$(diskutil info ${gCurrent_Partitions[k]} |grep -i Mounted |sed -e "s| Mounted:||" -e "s| ||g")
        if [[ ${gConfirm_Mounted} == *"Yes"* ]];
          then
            echo ${gCurrent_Partitions[k]} >> ${gMountPartition}
        fi
      done
      diskutil eject ${gDisk[i]}
  fi
done


#
# Fix RTLWlanUSB sleep problem credit B1anker & syscl/lighting/Yating Zhou. @PCBeta.
#

gRTWlan_kext=
gMAC_adr=$(ioreg -rc $gRTWlan_kext | sed -n "/IOMACAddress/ s/.*= <\(.*\)>.*/\1/ p")

if [[ "$gMAC_adr" != 0 ]];
  then
    gRT_Config="/Applications/Wireless Network Utility.app"/${gMAC_adr}rfoff.rtl

    if [ ! -f $gRT_Config ];
      then
        gRT_Config=$(ls "/Applications/Wireless Network Utility.app"/*rfoff.rtl)
    fi

    osascript -e 'quit app "Wireless Network Utility"'
    echo "1" > "$gRT_Config"
    open "/Applications/Wireless Network Utility.app"
fi
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
19:43 $ 
```

**Comment by OldManLink** [comment]
And this:
```
19:43 $ diskutil list
/dev/disk0 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *1.0 TB     disk0
   1:                        EFI EFI                     209.7 MB   disk0s1
   2:          Apple_CoreStorage Macintosh SSD           999.7 GB   disk0s2
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3

/dev/disk1 (internal, virtual):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:                  Apple_HFS Macintosh SSD          +999.3 GB   disk1
                                 Logical Volume on disk0s2
                                 F0CF6AA4-CACC-4F43-A027-547271C2E57E
                                 Unlocked Encrypted

✔ ~/CodeRoot/other/sleepFix [master L|✔] 
19:43 $ 
```

**Comment by syscl** [comment]
@OldManLink Hmm, the whole disk2 has disappeared, weird. How about sleep the laptop for a while then check if the disk2 reappear again?

**Comment by OldManLink** [comment]
OK, I will sleep the laptop for one minute and then wake it up again to see what happens.

**Comment by OldManLink** [comment]
Done: still no disk on the desktop, and the result of `diskutil list` appears to be the same:
```
19:44 $ diskutil list
/dev/disk0 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *1.0 TB     disk0
   1:                        EFI EFI                     209.7 MB   disk0s1
   2:          Apple_CoreStorage Macintosh SSD           999.7 GB   disk0s2
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3

/dev/disk1 (internal, virtual):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:                  Apple_HFS Macintosh SSD          +999.3 GB   disk1
                                 Logical Volume on disk0s2
                                 F0CF6AA4-CACC-4F43-A027-547271C2E57E
                                 Unlocked Encrypted

✔ ~/CodeRoot/other/sleepFix [master L|✔] 
19:48 $ 
```

**Comment by syscl** [comment]
So after executing ```sysclusbfix.sleep```, the whole disk2 disappear?

**Comment by OldManLink** [comment]
I only did what you suggested, i.e. sleep the laptop and wake it up a minute later. I have not executed any scripts yet after that.

**Comment by OldManLink** [comment]
If you like I could reboot the laptop and see if the disk comes back. Or if there's a better way, I'd love to know it!

**Comment by syscl** [comment]
@OldManLink Sounds good, thank you. I want to improve my program so that you can enjoy it as well!

**Comment by syscl** [comment]
Please let me know if reboot resolve the disk disappear issue.

**Comment by OldManLink** [comment]
Alright. I will reboot the laptop and I expect to answer here again in about five minutes.

**Comment by syscl** [comment]
No worry.

**Comment by OldManLink** [comment]
OK, I am back, and so is the SD disk. I await your next instructions!

**Comment by syscl** [comment]
OK, I update my Fix-usb-script a bit, would you mind give it a try, the latest version should get rid of the RTWlan issue if you don't have the related kext. And the code should be a bit different after you execute the fixUSB.sh, please let me know if the new sysclusbfix.sh change by the following
```sh
cat /etc/sysclusbfix.sleep
```

thank you.

**Comment by syscl** [comment]
We will then try to resolve the SD card eject issue.

**Comment by OldManLink** [comment]
I'll just put everything here in one block:
```
Last login: Wed Aug 29 19:58:16 on console
✔ ~ 
19:58 $ cd CodeRoot/other/sleepFix/
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
20:03 $ git remupp
Fetching origin
✔ ~/CodeRoot/other/sleepFix [master ↓·1|✔] 
20:03 $ git pull
Updating 1a0916a..7dc4273
Fast-forward
 fixUSB.sh | 4 ++++
 1 file changed, 4 insertions(+)
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
20:03 $ ./fixUSB.sh 
[  OK  ]  Generate configuration file of sleepwatcher launch daemon.
[  OK  ]  Generating script to unmount external devices before sleep (c) syscl/lighting/Yating Zhou.
[  OK  ]  Generate script to load RTWlanUSB upon sleep.
[ ---> ]  Installing external devices sleep patch...
Password:
[  OK  ]  Install sleepwatcher daemon.
[  OK  ]  Install configuration of sleepwatcher daemon.
[  OK  ]  Install sleep script.
[  OK  ]  Install wake script.
[  OK  ]  Fix the permissions of /etc/sysclusbfix.sleep.
[  OK  ]  Fix the permissions of /etc/sysclusbfix.wake.
[  OK  ]  Trigger startup service of syscl.usb.fix.
[  OK  ]  Clean up.
[ Note ]  DONE! Sleep to see change.
[ Note ]  Feel free to post issue on https://github.com/syscl/Fix-usb-sleep.
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
20:04 $ cat /etc/sysclusbfix.sleep
#!/bin/sh
#
# This script aims to unmount all external devices automatically before sleep.
#
# Without this procedure, various computers with OS X/Mac OS X(even on a real Mac) suffer from "Disk not ejected properly"
# issue when therere external devices plugged-in. Thats the reason why I created this script to fix this issue. (syscl/lighting/Yating Zhou)
#
# All credit to Bernhard Baehr (bernhard.baehr@gmx.de), without his great sleepwatcher dameon, this fix will not be created.
#

#
# Added unmount Disk for "OS X" (c) syscl/lighting/Yating Zhou.
#
gMountPartition="/tmp/com.syscl.externalfix"
gDisk=($(diskutil list |grep -i -i "dev" |grep -i -o "disk[0-9]"))

for ((i=0; i<${#gDisk[@]}; ++i))
do
  gProtocol=$(diskutil info ${gDisk[i]} |grep -i "Protocol" |sed -e "s|Protocol:||" -e "s| ||g")
  if [[ ${gProtocol} == *"USB"* ]];
    then
      gCurrent_Partitions=($(diskutil list ${gDisk[i]} |grep -o "disk[0-9]s[0-9]"))
      for ((k=0; k<${#gCurrent_Partitions[@]}; ++k))
      do
        gConfirm_Mounted=$(diskutil info ${gCurrent_Partitions[k]} |grep -i Mounted |sed -e "s| Mounted:||" -e "s| ||g")
        if [[ ${gConfirm_Mounted} == *"Yes"* ]];
          then
            echo ${gCurrent_Partitions[k]} >> ${gMountPartition}
        fi
      done
      diskutil eject ${gDisk[i]}
  fi
done

✔ ~/CodeRoot/other/sleepFix [master L|✔] 
20:04 $ 
```

**Comment by syscl** [comment]
Good, the code is better than before (I will reconstruct the RTLWan kext code block later on).

Would you please provide the log after executing the following commands
```sh
diskutil list
diskutil eject disk2
diskutil list
```

I just want to double check if the ```eject``` argument cause your SD card issue.

Thank you.

**Comment by OldManLink** [comment]
Here you go: 
```
20:08 $ diskutil list
/dev/disk0 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *1.0 TB     disk0
   1:                        EFI EFI                     209.7 MB   disk0s1
   2:          Apple_CoreStorage Macintosh SSD           999.7 GB   disk0s2
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3

/dev/disk1 (internal, virtual):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:                  Apple_HFS Macintosh SSD          +999.3 GB   disk1
                                 Logical Volume on disk0s2
                                 F0CF6AA4-CACC-4F43-A027-547271C2E57E
                                 Unlocked Encrypted

/dev/disk2 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *196.9 GB   disk2
   1:                        EFI EFI                     209.7 MB   disk2s1
   2:                  Apple_HFS Baseqi Ninja            196.5 GB   disk2s2

✔ ~/CodeRoot/other/sleepFix [master L|✔] 
20:09 $ diskutil eject disk2
Disk disk2 ejected
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
20:09 $ diskutil list
/dev/disk0 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *1.0 TB     disk0
   1:                        EFI EFI                     209.7 MB   disk0s1
   2:          Apple_CoreStorage Macintosh SSD           999.7 GB   disk0s2
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3

/dev/disk1 (internal, virtual):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:                  Apple_HFS Macintosh SSD          +999.3 GB   disk1
                                 Logical Volume on disk0s2
                                 F0CF6AA4-CACC-4F43-A027-547271C2E57E
                                 Unlocked Encrypted

✔ ~/CodeRoot/other/sleepFix [master L|✔] 
20:09 $ 
```

**Comment by syscl** [comment]
That's it, the diskutil eject cause the whole disk2 disappear. Hmm, why that would happen? Maybe we should use unmount command upon SD-card slot to resolve this issue. Would you give it a try after re-insert your sd-card then type in the following?
```sh
diskutil list
diskutil unmountDisk disk2
diskutil list
```

**Comment by OldManLink** [comment]
Here you go:
```
Last login: Wed Aug 29 20:14:28 on console
✔ ~ 
20:15 $ diskutil list
/dev/disk0 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *1.0 TB     disk0
   1:                        EFI EFI                     209.7 MB   disk0s1
   2:          Apple_CoreStorage Macintosh SSD           999.7 GB   disk0s2
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3

/dev/disk1 (internal, virtual):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:                  Apple_HFS Macintosh SSD          +999.3 GB   disk1
                                 Logical Volume on disk0s2
                                 F0CF6AA4-CACC-4F43-A027-547271C2E57E
                                 Unlocked Encrypted

/dev/disk2 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *196.9 GB   disk2
   1:                        EFI EFI                     209.7 MB   disk2s1
   2:                  Apple_HFS Baseqi Ninja            196.5 GB   disk2s2

✔ ~ 
20:16 $ diskutil unmountDisk disk2
Unmount of all volumes on disk2 was successful
✔ ~ 
20:16 $ diskutil list
/dev/disk0 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *1.0 TB     disk0
   1:                        EFI EFI                     209.7 MB   disk0s1
   2:          Apple_CoreStorage Macintosh SSD           999.7 GB   disk0s2
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3

/dev/disk1 (internal, virtual):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:                  Apple_HFS Macintosh SSD          +999.3 GB   disk1
                                 Logical Volume on disk0s2
                                 F0CF6AA4-CACC-4F43-A027-547271C2E57E
                                 Unlocked Encrypted

/dev/disk2 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *196.9 GB   disk2
   1:                        EFI EFI                     209.7 MB   disk2s1
   2:                  Apple_HFS Baseqi Ninja            196.5 GB   disk2s2

✔ ~ 
20:16 $ 
```

**Comment by syscl** [comment]
Cool, give me a few minutes to improve it.

syscl

**Comment by syscl** [comment]
Please give the latest fixUSB.sh a try.

Thank you.

**Comment by OldManLink** [comment]
Looks like an "oops"
```
20:16 $ 
✔ ~ 
20:32 $ pwd
/Users/peter
✔ ~ 
20:32 $ cd CodeRoot/other/sleepFix/
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
20:33 $ git remupp
Fetching origin
✔ ~/CodeRoot/other/sleepFix [master ↓·1|✔] 
20:33 $ git pull
Updating 7dc4273..ee1d78a
Fast-forward
 fixUSB.sh | 132 +++++++++++++++++++++++++++++++++++++++++++++++---------------------------------------------
 1 file changed, 67 insertions(+), 65 deletions(-)
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
20:33 $ ./fixUSB.sh 
./fixUSB.sh: line 260: syntax error near unexpected token `('
./fixUSB.sh: line 260: `    echo '# Added mount Disk for "OS X" (c) syscl/lighting/Yating Zhou.'                                                                                    >> "$gUSBWakeScript"'
✘-2 ~/CodeRoot/other/sleepFix [master L|✔] 
20:33 $
```

**Comment by syscl** [comment]
Thank you for your report, please try the latest one.

**Comment by OldManLink** [comment]
Here you go: 
```
20:40 $ ./fixUSB.sh 
[  OK  ]  Generate configuration file of sleepwatcher launch daemon.
[  OK  ]  Generating script to unmount external devices before sleep (c) syscl/lighting/Yating Zhou.
[  OK  ]  Generate script to load RTWlanUSB upon sleep.
[ ---> ]  Installing external devices sleep patch...
Password:
[  OK  ]  Install sleepwatcher daemon.
[  OK  ]  Install configuration of sleepwatcher daemon.
[FAILED]  Install sleep script.
cp: /tmp/sysclusbfix.sleep: No such file or directory
[  OK  ]  Install wake script.
[  OK  ]  Fix the permissions of /etc/sysclusbfix.sleep.
[  OK  ]  Fix the permissions of /etc/sysclusbfix.wake.
[  OK  ]  Trigger startup service of syscl.usb.fix.
[FAILED]  Clean up.
rm: /tmp/sysclusbfix.sleep: No such file or directory
[ Note ]  DONE! Sleep to see change.
[ Note ]  Feel free to post issue on https://github.com/syscl/Fix-usb-sleep.
✘-1 ~/CodeRoot/other/sleepFix [master L|✔] 
20:40 $ 
```

**Comment by OldManLink** [comment]
Now I must drive home, but I will check my mail again in about 30 minutes.

**Comment by syscl** [comment]
See you.

**Comment by OldManLink** [comment]
Ok, I am home now, so will grab some food then try out the new version.

**Comment by OldManLink** [comment]
Here we go:
```
21:45 $ cd CodeRoot/other/sleepFix/
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
21:45 $ git remupp
Fetching origin
✔ ~/CodeRoot/other/sleepFix [master ↓·1|✔] 
21:45 $ git pull
Updating f587f8e..e2486c2
Fast-forward
 fixUSB.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
21:45 $ ./fixUSB.sh 
[  OK  ]  Generate configuration file of sleepwatcher launch daemon.
[  OK  ]  Generate script to load RTWlanUSB upon sleep.
[ ---> ]  Installing external devices sleep patch...
Password:
[  OK  ]  Install sleepwatcher daemon.
[  OK  ]  Install configuration of sleepwatcher daemon.
[  OK  ]  Install sleep script.
[  OK  ]  Install wake script.
[  OK  ]  Fix the permissions of /etc/sysclusbfix.sleep.
[  OK  ]  Fix the permissions of /etc/sysclusbfix.wake.
[  OK  ]  Trigger startup service of syscl.usb.fix.
[  OK  ]  Clean up.
[ Note ]  DONE! Sleep to see change.
[ Note ]  Feel free to post issue on https://github.com/syscl/Fix-usb-sleep.
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
21:46 $ 
``` 
Next I will sleep the laptop for one minute and report what happens.

**Comment by OldManLink** [comment]
```
21:46 $ echo "Sleep done"
Sleep done
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
21:51 $ diskutil list
/dev/disk0 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *1.0 TB     disk0
   1:                        EFI EFI                     209.7 MB   disk0s1
   2:          Apple_CoreStorage Macintosh SSD           999.7 GB   disk0s2
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3

/dev/disk1 (internal, virtual):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:                  Apple_HFS Macintosh SSD          +999.3 GB   disk1
                                 Logical Volume on disk0s2
                                 F0CF6AA4-CACC-4F43-A027-547271C2E57E
                                 Unlocked Encrypted

✔ ~/CodeRoot/other/sleepFix [master L|✔] 
21:52 $ 
```

**Comment by OldManLink** [comment]
So now the disk is gone again after the sleep.

**Comment by syscl** [comment]
OK. Would you please re-insert the sd card then try to execute the script manually
```sh
sudo /etc/sysclusbfix.sleep
```

Thank you.

**Comment by OldManLink** [comment]
Now it's getting late, so I'll call it a day and check in again tomorrow. 

Thanks for trying to fix this!

**Comment by syscl** [comment]
Have a good night.

It seems that the sleep script has correctly unmount the sd card. How about executing the wake script(by ```sudo /etc/sysclusbfix.wake``` after wake from sleep? I think the problem now turns out to be the wake script.

**Comment by syscl** [comment]
TL;DR. It seems the SD card hasn't been ejected correctly which is operated by the *HCI kext. It seems the comment unmount then remount method does not work in this case. I wish I can come up to a better solution and will let you know. once I push a new update.

**Comment by syscl** [comment]
Please try the latest version of script.

**Comment by OldManLink** [comment]
Sorry for the delay, I was travelling without my MacBook for the last few days. So here is the latest console log: 
```
11:25 $ ./fixUSB.sh -u 
[  OK  ]  Unload /Library/LaunchDaemons/com.syscl.externalfix.sleepwatcher.plist.
[  OK  ]  Remove /usr/local/sbin/.
[  OK  ]  Remove /Library/LaunchDaemons/com.syscl.externalfix.sleepwatcher.plist.
[  OK  ]  Remove /etc/sysclusbfix.sleep.
[  OK  ]  Remove /etc/sysclusbfix.wake.
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
11:26 $ diskutil list
/dev/disk0 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *1.0 TB     disk0
   1:                        EFI EFI                     209.7 MB   disk0s1
   2:          Apple_CoreStorage Macintosh SSD           999.7 GB   disk0s2
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3

/dev/disk1 (internal, virtual):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:                  Apple_HFS Macintosh SSD          +999.3 GB   disk1
                                 Logical Volume on disk0s2
                                 F0CF6AA4-CACC-4F43-A027-547271C2E57E
                                 Unlocked Encrypted

/dev/disk2 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *196.9 GB   disk2
   1:                        EFI EFI                     209.7 MB   disk2s1
   2:                  Apple_HFS Baseqi Ninja            196.5 GB   disk2s2

✔ ~/CodeRoot/other/sleepFix [master L|✔] 
11:26 $ ./fixUSB.sh
[  OK  ]  Generate configuration file of sleepwatcher launch daemon.
[  OK  ]  Generate script to load RTWlanUSB upon sleep.
[ ---> ]  Installing external devices sleep patch...
[  OK  ]  Create /usr/local/sbin/.
[  OK  ]  Install sleepwatcher daemon.
[  OK  ]  Install configuration of sleepwatcher daemon.
[  OK  ]  Install sleep script.
[  OK  ]  Install wake script.
[  OK  ]  Fix the permissions of /etc/sysclusbfix.sleep.
[  OK  ]  Fix the permissions of /etc/sysclusbfix.wake.
[  OK  ]  Trigger startup service of syscl.usb.fix.
[  OK  ]  Clean up.
[ Note ]  DONE! Sleep to see change.
[ Note ]  Feel free to post issue on https://github.com/syscl/Fix-usb-sleep.
✔ ~/CodeRoot/other/sleepFix [master L|✔] 
11:27 $ cat /etc/sysclusbfix.sleep
#!/bin/sh
#
# This script aims to unmount all external devices automatically before sleep.
#
# Without this procedure, various computers with OS X/Mac OS X(even on a real Mac) suffer from "Disk not ejected properly"
# issue when therere external devices plugged-in. Thats the reason why I created this script to fix this issue. (syscl/lighting/Yating Zhou)
#
# All credit to Bernhard Baehr (bernhard.baehr@gmx.de), without his great sleepwatcher dameon, this fix will not be created.
#

#
# Added unmount Disk for "OS X" (c) syscl/lighting/Yating Zhou.
#
gMountPartition="/tmp/com.syscl.externalfix"
gDisk=($(diskutil list |grep -i -i "dev" |grep -i -o "disk[0-9]"))
gUSBDisk=""

for ((i=0; i<${#gDisk[@]}; ++i))
do
  gProtocol=$(diskutil info ${gDisk[i]} |grep -i "Protocol" |sed -e "s|Protocol:||" -e "s| ||g")
  is_SDCard=$(diskutil info ${gDisk[i]} |grep -i "Device / Media Name" |grep -i "SD Card")
  if [[ ${gProtocol} == *"USB"* ]];
    then
      gCurrent_Partitions=($(diskutil list ${gDisk[i]} |grep -o "disk[0-9]s[0-9]"))
      for ((k=0; k<${#gCurrent_Partitions[@]}; ++k))
      do
        gConfirm_Mounted=$(diskutil info ${gCurrent_Partitions[k]} |grep -i Mounted |sed -e "s| Mounted:||" -e "s| ||g")
        if [[ ${gConfirm_Mounted} == *"Yes"* ]];
          then
            echo ${gCurrent_Partitions[k]} >> ${gMountPartition}
        fi
      done
      if [ -z ${is_SDCard} ]; then
          diskutil eject ${gDisk[i]}
      else
          gUSBDisk="${gUSBDisk}
${gDisk[i]}"
      fi
  fi
done
echo -e "${gUSBDisk}
" | xargs -I {} diskutil eject {}

✔ ~/CodeRoot/other/sleepFix [master L|✔] 
11:28 $ sudo /etc/sysclusbfix.sleep
/etc/sysclusbfix.sleep: line 33: [: too many arguments
Unable to find disk for -e disk2
✘-1 ~/CodeRoot/other/sleepFix [master L|✔] 
11:30 $ 

```

**Comment by OldManLink** [comment]
Note: the disk is still mounted: 
```
11:33 $ diskutil list
/dev/disk0 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *1.0 TB     disk0
   1:                        EFI EFI                     209.7 MB   disk0s1
   2:          Apple_CoreStorage Macintosh SSD           999.7 GB   disk0s2
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3

/dev/disk1 (internal, virtual):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:                  Apple_HFS Macintosh SSD          +999.3 GB   disk1
                                 Logical Volume on disk0s2
                                 F0CF6AA4-CACC-4F43-A027-547271C2E57E
                                 Unlocked Encrypted

/dev/disk2 (internal, physical):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        *196.9 GB   disk2
   1:                        EFI EFI                     209.7 MB   disk2s1
   2:                  Apple_HFS Baseqi Ninja            196.5 GB   disk2s2

✔ ~/CodeRoot/other/sleepFix [master L|✔] 
11:33 $ 
```

**Comment by axeII** [comment]
Thank you for working on this. Any update? 🙏

**Comment by thomhastings** [comment]
Thank you for working on this. Any update? 🙏
